qiskit/transpiler/synthesis/aqc/cnot_unit_objective.py

```diff
@@ -0,0 +1,299 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+A definition of the approximate circuit compilation optimization problem based on CNOT unit
+definition.
+"""
+from abc import abstractmethod
+
+import numpy as np
+from numpy import linalg as la
+
+from .approximate import ApproximatingObjective
+from .elementary_operations import ry_matrix, rz_matrix, place_unitary, place_cnot, rx_matrix
+
+
+class CNOTUnitObjective(ApproximatingObjective):
+    """
+    A base class for a problem definition based on CNOT unit. This class may have different
+    subclasses for objective and gradient computations.
+    """
+
+    def __init__(self, num_qubits: int, cnots: np.ndarray) -> None:
+        """
+        Args:
+            num_qubits: number of qubits.
+            cnots: a CNOT structure to be used in the optimization procedure.
+        """
+        super().__init__()
+        self._num_qubits = num_qubits
+        self._cnots = cnots
+        self._num_cnots = cnots.shape[1]
+        # must be defined later
+        self._target_matrix = None
+
+    @abstractmethod
+    def objective(self, param_values: np.ndarray) -> float:
+        raise NotImplementedError
+
+    @abstractmethod
+    def gradient(self, param_values: np.ndarray) -> np.ndarray:
+        raise NotImplementedError
```

Comment:
I might be wrong here, but I dont thing you need these methods.
```suggestion
```

Reply:
Agree, removed, class marked as `ABC`.